src/main/scala/viper/gobra/frontend/Desugar.scala

```diff
@@ -2491,15 +2542,37 @@ object Desugar {
 
     }
 
-    def permissionD(ctx: FunctionContext)(perm: PPermission): Writer[in.Permission] = {
-      val src: Meta = meta(perm)
+    def permissionD(ctx: FunctionContext)(exp: PExpression): Writer[in.Expr] = {
+      maybePermissionD(ctx)(exp) getOrElse exprD(ctx)(exp)
+    }
+
+    def maybePermissionD(ctx: FunctionContext)(exp: PExpression): Option[Writer[in.Expr]] = {
+      val src: Meta = meta(exp)
       def goE(e: PExpression): Writer[in.Expr] = exprD(ctx)(e)
 
-      perm match {
-        case PFullPerm() => unit(in.FullPerm(src))
-        case PNoPerm() => unit(in.NoPerm(src))
-        case PFractionalPerm(left, right) => for { l <- goE(left); r <- goE(right) } yield in.FractionalPerm(l, r)(src)
-        case PWildcardPerm() => unit(in.WildcardPerm(src))
+      exp match {
+        case n: PInvoke => info.resolve(n) match {
+          case Some(_: ap.Conversion) =>
+            Some(for {
+              // the well-definedness checker ensures that there is exactly one argument
+              arg <- permissionD(ctx)(n.args.head)
+            } yield in.Conversion(in.PermissionT(Addressability.conversionResult), arg)(src))
+          case _ => None
+        }
```

Comment:
I also meant that I do not understand why this case is required at all?

Reply:
could you try removing this case and then trying the case, I think it would still work.

Reply:
I actually tried it without the special case and it crashed
```
[error] Exception in thread "main" java.util.concurrent.ExecutionException: java.util.concurrent.ExecutionException: java.lang.AssertionError: assertion failed: Expected both operands to be of the same sort, but found Perm (Z) and Int (1 / 2).
[error] 	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
[error] 	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
[error] 	at viper.silicon.Silicon.verify(Silicon.scala:199)
[error] 	at viper.silicon.Silicon.verify(Silicon.scala:157)
[error] 	at viper.gobra.backend.Silicon.$anonfun$verify$1(Silicon.scala:27)
[error] 	at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:672)
[error] 	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:431)
[error] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[error] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[error] 	at java.base/java.lang.Thread.run(Thread.java:834)
[error] Caused by: java.util.concurrent.ExecutionException: java.lang.AssertionError: assertion failed: Expected both operands to be of the same sort, but found Perm (Z) and Int (1 / 2).
[error] 	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
[error] 	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
[error] 	at viper.silicon.verifier.DefaultMasterVerifier.$anonfun$verify$9(DefaultMasterVerifier.scala:240)
[error] 	at scala.collection.immutable.List.flatMap(List.scala:293)
[error] 	at scala.collection.immutable.List.flatMap(List.scala:79)
[error] 	at viper.silicon.verifier.DefaultMasterVerifier.verify(DefaultMasterVerifier.scala:240)
[error] 	at viper.silicon.Silicon.viper$silicon$Silicon$$runVerifier(Silicon.scala:235)
[error] 	at viper.silicon.Silicon$$anon$1.call(Silicon.scala:193)
[error] 	at viper.silicon.Silicon$$anon$1.call(Silicon.scala:192)
[error] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] 	... 3 more
[error] Caused by: java.lang.AssertionError: assertion failed: Expected both operands to be of the same sort, but found Perm (Z) and Int (1 / 2).
[error] 	at scala.Predef$.assert(Predef.scala:280)
```